File: src/utils/math/BoundedFunction.java
package utils.math;

import java.util.List;

public interface BoundedFunction<T> extends Function<T> {
    double upperBound();
    double lowerBound();

    default T getUpper(){
        return get(upperBound());
    }
    default T getLower(){
        return get(lowerBound());
    }

    BoundedFunction<T> offset(double offset) throws ClassCastException;
}

File: src/utils/math/BoundedFunction.kt
package utils.math

interface BoundedFunction<T> : (Double) -> T {
    fun upperBound(): Double
    fun lowerBound(): Double

    val upper: T
        get() = this(upperBound())
    val lower: T
        get() = this(lowerBound())

    fun bounded(input: Double) : T = this(input.coerceIn(lowerBound()..upperBound()))

    fun offset(offset: Double): BoundedFunction<T>
}